[JSC] Use JavaScriptCore framework dyld UUID for computing bytecode cache version

hyjorc1 opened the pull request:
#### eff5765b8477190745b7957c20f46e56c6e89fb5
<pre>
[JSC] Use JavaScriptCore framework dyld UUID for computing bytecode cache version
<a href="https://bugs.webkit.org/show_bug.cgi?id=272660">https://bugs.webkit.org/show_bug.cgi?id=272660</a>
<a href="https://rdar.apple.com/126195542">rdar://126195542</a>

Reviewed by Yusuke Suzuki.

Previously, we used the JSCBuiltins.o, CachedTypes.o, and project source version
number for computing JSC bytecode cache version. That may not strong enough
to reject a stale bytecode cache. This patch utilizes JavaScriptCore.framework’s
UUID to compute the bytecode cache hash version for Darwin OSes and uses
__TIMESTAMP__ for the others.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::GenericCacheEntry::GenericCacheEntry):
(JSC::GenericCacheEntry::isUpToDate const):
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp: Added.
(JSC::computeJSCBytecodeCacheVersion):
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in: Removed.
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.h:

Canonical link: <a href="https://commits.webkit.org/277548@main">https://commits.webkit.org/277548@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69b7fb10d6c76645b166c605cbeaa76a4b4b7412

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47899 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27100 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50741 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50580 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43952 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24585 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38964 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24758 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41412 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20264 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22236 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/42594 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5946 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41188 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44244 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43028 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52474 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47390 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19279 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46271 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24211 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45314 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25001 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54888 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6797 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23932 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11280 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp:
```diff
@@ -0,0 +1,75 @@
+
+std::optional<uint32_t> computeJSCBytecodeCacheVersion()
+{
+    static std::optional<uint32_t> cacheVersion;
```
Let's use `LazyNeverDestroyed<uint32_t>` and `std::call_once`.

Constellation commented on Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp:
```diff
@@ -0,0 +1,75 @@
+            dataLogLnIf(JSCBytecodeCacheVersionInternal::verbose, "i:", i, " imageName:", imageName);
+            const struct mach_header* header = _dyld_get_image_header(i);
+
+            if (header && _dyld_get_image_uuid(header, uuid)) {
```
I don't think you need to search for the current framework.
I think, `_dyld_get_dlopen_image_header(RTLD_SELF)` should work (please double-check by looking at UUID v.s. framework UUID, and review dyld code).

Constellation replied:
Hmm, but let's be a bit careful about this. Using `HAVE(DYLD_DLOPEN_IMAGE_HEADER_SPI)` is probably required.

Constellation replied:
Or, I wonder if we should just use `dyld_image_header_containing_address`.

1. Wrap two functions, each one has `NEVER_INLINE` and `JS_EXPORT_PRIVATE`
2. Both should use NO_TAIL_CALLS to suppress tail calls 
3. And in the nested function, using `__builtin_return_address(0)` to obtain the outer function's address. Then this address is 100% JSC framework's address
4. Pass it to `dyld_image_header_containing_address`

kmiller68 replied:
Couldn't we just call `dyld_image_header_containing_address(computeJSCBytecodeCacheVersion)`? Even if it gets inlined that function should still by in the dylib, IIUC.

Constellation commented on Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in:
```diff
@@ -1,7 +1,71 @@
+    std::call_once(onceFlag, [] {
+        uuid_t uuid;
+        uuid_string_t uuidString = { };
+        void* jsFunctionAddr = bitwise_cast<void*>(&JSEvaluateScript);
```
Probably using `computeJSCBytecodeCacheVersion` would be the best.

Constellation commented on Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in:
```diff
@@ -1,7 +1,71 @@
+static constexpr bool verbose = false;
+}
+
+uint32_t computeJSCBytecodeCacheVersion()
```
And let's put `NEVER_INLINE` for extra safety.

Constellation commented on Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in:
```diff
@@ -1,7 +1,71 @@
+        uuid_t uuid;
+        uuid_string_t uuidString = { };
+        void* jsFunctionAddr = bitwise_cast<void*>(&JSEvaluateScript);
+        if (const mach_header* header = dyld_image_header_containing_address(jsFunctionAddr); header && _dyld_get_image_uuid(header, uuid)) {
```
Add definition of `dyld_image_header_containing_address` to `dyldSPI.h`.

Constellation commented on Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in:
```diff
@@ -1,7 +1,75 @@
 #include <wtf/text/SuperFastHash.h>
 
-const uint32_t JSCBytecodeCacheVersion = SuperFastHash::computeHash(CACHED_TYPES_CKSUM);
+#if OS(DARWIN)
```
I think we do not need to generate these things on Darwin at all.
Can we remove generating code from Xcode and use normal C++ code instead of `cpp.in`?
We should have `CACHED_TYPES_CKSUM` etc. things only for non Darwin builds.

Constellation commented on Source/JavaScriptCore/CMakeLists.txt:
```diff
@@ -151,31 +151,6 @@ set(JavaScriptCore_OBJECT_LUT_SOURCES
-    runtime/CachedTypes.cpp
-    ${JavaScriptCore_DERIVED_SOURCES_DIR}/JSCBytecodeCacheVersion.cpp
-    ${JavaScriptCore_DERIVED_SOURCES_DIR}/JSCBuiltins.cpp
-)
```
This is removing `CachedTypes.cpp` and `JSCBuiltins.cpp` too. They are required.

Constellation commented on Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
```diff
@@ -13186,6 +13169,7 @@
 			buildActionMask = 2147483647;
 			files = (
 				44F93E0E2AE71FBD00FFA37C /* JavaScriptCoreFramework.cpp in Sources */,
+				FFF3372F2BCDE4A40070D04B /* JSCBytecodeCacheVersion.cpp in Sources */,
```
This looks very strange. I think probably this is not a right place.

Constellation replied:
I recommend first completely removing  `JSCBytecodeCacheVersion.cpp` etc. from Xcode, and just re-add them as a normal files.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Use JavaScriptCore framework dyld UUID for computing bytecode cache version
https://bugs.webkit.org/show_bug.cgi?id=272660
rdar://126195542

Reviewed by Yusuke Suzuki.

Previously, we used the JSCBuiltins.o, CachedTypes.o, and project source version
number for computing JSC bytecode cache version. That may not strong enough
to reject a stale bytecode cache. This patch utilizes JavaScriptCore.framework’s
UUID to compute the bytecode cache hash version for Darwin OSes and uses
__TIMESTAMP__ for the others.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::GenericCacheEntry::GenericCacheEntry):
(JSC::GenericCacheEntry::isUpToDate const):
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp: Added.
(JSC::computeJSCBytecodeCacheVersion):
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp.in: Removed.
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.h:

Canonical link: https://commits.webkit.org/277548@main
